### PR TITLE
Bootcamp

### PR DESCRIPTION
**Pull request by LinghanX**
Created CRUD actions for Bootcamp model
Created test for Boocamp model

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,11 @@
 ![Build Status](https://codeship.com/projects/6d205b90-42b3-0133-8114-0a794f3732af/status?branch=master)![Code Climate](https://codeclimate.com/github/MkNiz/Bootcamp-Reviews.png) ![Coverage Status](https://coveralls.io/repos/MkNiz/Bootcamp-Reviews/badge.png)
+
+
+ Bootcamp
```
Comment (dreamingfurther) [inline README.md]: You should update your README for the next developer using the site or as a guide of the features on the site.

File: spec/features/bootcamp_new_features_spec.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+} do
+  scenario 'user visit new page' do
+    visit bootcamps_path
+    click_link 'New bootcamp'
+    expect(page).to have_content('Create your bootcamp')
+  end
+
+  scenario 'user submtits a new form' do
+    visit new_bootcamp_path
+    bootcamp = FactoryGirl.create(:bootcamp)
```
Comment (dreamingfurther) [inline spec/features/bootcamp_new_features_spec.rb]: what happens if the Factory somehow created this object with `nil` values for all the fields?

In that case your test will be be asserting that the page has the value `nil` on it (or nothing) and will pass but you will actually still have a broken application.

For testing its actually a better pattern to hard code the strings for the bootcamp name, city, description, etc rather than calling those from a factory.

Also if you are creating a bootcamp by using the form on the page what is the value to creating a bootcamp through the use of a `Factory`?

File: spec/features/user_visits_index_pages_spec.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# require 'rails_helper'
```
Comment (dreamingfurther) [inline spec/features/user_visits_index_pages_spec.rb]: Either use these tests or delete the file. Do not commit files that are full of commented out code. its messy.

**Comment by dreamingfurther** [comment]
@LinghanX you also need to merge master into this branch to resolve these conflicts.

File: Gemfile [hunk 4]
```diff
@@ -45,7 +46,7 @@ end
 
 group :test do
   gem 'coveralls', require: false
-  gem 'shoulda-matchers'
+  gem 'faker'
```
Comment (cmkoller) [inline Gemfile]: Why did you delete the `shoulda-matchers` gem? Isn't that used in other tests? If not, it should be!

Comment (cmkoller) [inline Gemfile]: Oh never mind, I see it's also listed below.

File: app/controllers/bootcamps_controller.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    params.require(:bootcamp).permit(
+    :name,
+    :city,
+    :description, 
+    :country,
+    :rating,
+    :website,
+    :price,
+    :duration,
+    :is_online)
```
Comment (cmkoller) [inline app/controllers/bootcamps_controller.rb]: All of these symbols should be indented by one, with the closing parentheses on a new line:

```
code code code(
  :stuff, 
  :more_stuff, 
  :even_more_stuff
)
```

File: app/controllers/bootcamps_controller.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      render :new
+    end
+  end
+
+  def destroy
+    @bootcamp = Bootcamp.find(params[:id])
+    @bootcamp.destroy
+    flash[:success] = 'Bootcamp deleted'
+    redirect_to bootcamps_path
+  end
```
Comment (cmkoller) [inline app/controllers/bootcamps_controller.rb]: Holy wow this is a lot of actions in one PR! It would probably be better in the future to split this work up into separate PRs, each of these is kind of its own feature. (Well, `edit`/`update` are paired and `new`/`create` are paired, but everything else can be its own separate feature.)

File: app/views/bootcamps/_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,42 @@
+<%= form_for @bootcamp do |f| %>
+
+  <%= f.label :name %>
+  <%= f.text_field :name, placeholder: 'Launch Academy' %>
+  <br>
```
Comment (cmkoller) [inline app/views/bootcamps/_form.html.erb]: You shouldn't be using `br` here! You're only using it to get more space in between your elements, and that's something that should be done using CSS

File: app/views/bootcamps/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,20 @@
+<h1>List of Bootcamps</h1>
+<% @bootcamps.each do |bootcamp| %>
+<table>
```
Comment (cmkoller) [inline app/views/bootcamps/index.html.erb]: Lines 3 to 17 should be indented by one

File: app/views/bootcamps/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  <thead>
+    <tr>
+      <th width="200"><%= link_to bootcamp.name, bootcamp_path(bootcamp) %></th>
```
Comment (cmkoller) [inline app/views/bootcamps/index.html.erb]: I wouldn't recommend putting an explicit width on a table cell. Foundation tables are usually pretty good at looking nice on their own, and this makes your table very un-dynamic

File: app/views/bootcamps/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+      <th width="150">
+        <ul>
+          <li>Rating: <%= @bootcamp.rating %></li>
+          <li>City: <%= @bootcamp.city %></li>
+          <li>Price: <%= number_to_currency @bootcamp.price, precision: 0 %></li>
+          <li><%= link_to 'Website', @bootcamp.website %></li>
+          <li>Duration: <%= @bootcamp.duration %></li>
+          <li>Category: <%= @bootcamp.is_online ? 'Online' : 'Onsite' %>
+          </li>
+        </ul>
```
Comment (cmkoller) [inline app/views/bootcamps/show.html.erb]: What are you doing with a `ul` inside a table? That's not very traditional. Either use a `ul` or fully embrace the table structure!

Looking at this more, I don't think a table is the right call at all here. I'd need to see the actual page itself to be sure, but it looks like you're using this table purely for formatting (rather than actually displaying lots of data in a table-like format) - in which case you should be using something like Foundation's grid system instead.

File: spec/features/bootcamp_delete_features_spec.rb [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+  So that I can delete a bootcamp that it's not longer offer.
+} do
+  scenario 'user visit edit page' do
+    bootcamp = FactoryGirl.create(:bootcamp)
+
+    visit edit_bootcamp_path(bootcamp)
+    click_link 'Delete'
+
+    expect(page).to_not have_content(bootcamp.name)
+  end
```
Comment (cmkoller) [inline spec/features/bootcamp_delete_features_spec.rb]: It looks like right now any user is able to delete a bootcamp. I expect a new PR changing that fact once you have the admin functionality merged in - only admins and the user who created this bootcamp should be allowed to delete it!

Comment (cmkoller) [inline spec/features/bootcamp_delete_features_spec.rb]: Actually, you can do half of that right now - you should restrict a user's ability to delete something such that ONLY the user who created the bootcamp can delete it

File: spec/features/bootcamp_edit_features_spec.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    expect(page).to have_content('Edit')
+  end
+
+  scenario 'user introduces wrong data' do
+    bootcamp = FactoryGirl.create(:bootcamp)
+    visit edit_bootcamp_path(bootcamp)
+    fill_in('Rating', with: 'six')
+    click_button 'Submit'
+    expect(page).to have_content('Something went wrong')
+  end
```
Comment (cmkoller) [inline spec/features/bootcamp_edit_features_spec.rb]: Same here as above - ONLY the user who created the bootcamp should be able to edit it.

File: spec/features/bootcamp_index_features_spec.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+feature 'user view index page of bootcamp', %{
+  As a signed in user,
+  I want to be able to read all the information about bootcamps,
+  So that I can get an idea of what bootcamps are suitable to me.
+} do
+  scenario 'user visit index page' do
+    bootcamp = FactoryGirl.create(:bootcamp)
+    bootcamp2 = FactoryGirl.create(:bootcamp)
+    bootcamp3 = FactoryGirl.create(:bootcamp)
+    bootcamp4 = FactoryGirl.create(:bootcamp)
```
Comment (cmkoller) [inline spec/features/bootcamp_index_features_spec.rb]: Listen to Hound here! You have a bunch of variables defined that you never use!

Please try to take care of Hound errors like this before sending the PR to me for review, it'll save time

File: spec/features/bootcamp_new_features_spec.rb [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+  I want to be able to create a new bootcamp,
+  So that I can publish my bootcamps and reviews
+} do
+  scenario 'user visit new page' do
+    visit bootcamps_path
+    click_link 'New bootcamp'
+    expect(page).to have_content('Create your bootcamp')
+  end
+
+  scenario 'user submtits a new form' do
```
Comment (cmkoller) [inline spec/features/bootcamp_new_features_spec.rb]: not that it realllly matters but there's a typo here ^ :)

File: spec/features/bootcamp_new_features_spec.rb [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    fill_in('City', with: 'Phoenix')
+    fill_in('Description', with: 'After my time in LA, i grow deep fear towards Hound')
+    fill_in('Country', with: 'United States')
+    fill_in('Rating', with: 8)
+    fill_in('Website', with: 'http://www.launchacademy.com')
+    fill_in('Price', with: 8_973_949_528_943_985_000_000)
+    fill_in('Duration', with: 10)
+    choose('bootcamp_is_online_true')
+    click_button('Submit')
+    expect(page).to have_content('Launch Academy')
```
Comment (cmkoller) [inline spec/features/bootcamp_new_features_spec.rb]: Could we have another expectation about the page having some sort of content like "Your bootcamp was successfully submitted"? This should be displayed via a flash message in the controller action

File: spec/models/user_spec.rb [hunk 16]
```diff
@@ -1,5 +1,4 @@
 require 'rails_helper'
 
 RSpec.describe User, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
 end
```
Comment (cmkoller) [inline spec/models/user_spec.rb]: Does the user have many bootcamps? If so, that's a relationship you should put in this test instead of just deleting that line!

File: app/controllers/bootcamps_controller.rb [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+  def new
+    @bootcamp = Bootcamp.new
+  end
+
+  def create
+    @bootcamp = Bootcamp.new(bootcamp_params)
+
+    if @bootcamp.save
+      redirect_to bootcamp_path(@bootcamp)
```
Comment (cmkoller) [inline app/controllers/bootcamps_controller.rb]: Flash a success message if it's successful!

**Comment by LinghanX** [comment]
Fixed some problems, major changes includes:

1. Indent issues now should be eliminated.
2. Use foundation grid system instead of tables
3. change redundancy in spec files.
